feat!: add an eof function

amaanq opened the pull request:
Closes #160 

Instead of relying on the null char `'\0'` to match the end of a file (where that can also match any null char inside the file as well), one can now do this:

```js
module.exports = grammar({
    name: 'test',

    rules: {
        source_file: $ => seq('hello', eof()),
    }
});
```

ahlinc commented on lib/src/parser.c:
```diff
@@ -482,7 +482,7 @@ static Subtree ts_parser__lex(
     );
     ts_lexer_start(&self->lexer);
-    bool found_token = self->language->lex_fn(&self->lexer.data, lex_mode.lex_state);
+    bool found_token = self->language->lex_fn(&self->lexer.data, lex_mode.lex_state, &self->lexer.accepted_eof);
```
I'm not sure that such approach for the implementation is correct, it looks like you created a parallel set of eof states to lex_states and their intersection in a generated parser.c. I thought implementation of the eof() would just generate additional lex states by using existing mechanism, so for some rules the would be generated aditional eof checks and on possitive values transitions to an  aditional state with the aux_sym_eof acceptance.

amaanq replied:
Not exactly, another solution can be to use the field in the TSLexer instead of Lexer, but since that's exposed in the scanner it's maybe better not to.

The issue is we need to know if `aux_sym_eof` was already accepted once before (without advancing), and if so, then the builtin eof sym can be accepted

maxbrunsfeld replied:
I don't think it should matter whether or not the EOF token was accepted already. Conceptually, it seems like multiple layers of rules should be able to match on `EOF`. Is this intended for preventing infinite loops?

amaanq replied:
Yeah it would endlessly accept `aux_sym_eof` without this check and the parsing would never end.

maxbrunsfeld commented on cli/src/generate/nfa.rs:
```diff
@@ -445,27 +448,28 @@ impl<'a> NfaCursor<'a> {
     }
 
-    fn raw_transitions(&self) -> impl Iterator<Item = (&CharacterSet, bool, i32, u32)> {
+    fn raw_transitions(&self) -> impl Iterator<Item = (&CharacterSet, bool, i32, u32, bool)> {
```
I don't think `eof` should be modeled as part of an `Advance` action. I would think that it would just be a property of an `AcceptToken` action: "this token can only be accepted in this state if we are at EOF".

EOF seems different than a lookahead character set in that it's not something in the input stream that we *consume*; it's just a boolean property (are we or are we not at EOF) of a point in the input stream.

amaanq replied:
I was a bit unsure about how to model the eof action, yeah, there's no actual consumption of a character but we would still want to "advance" to the state when an eof token was used I'd think. Or, we can just accept in place in the zeroth state, is that what you were thinking? Here's my interpretation of it just being an `AcceptToken` action

```c
if (!*accepted_eof && eof) {
        *accepted_eof = true;
        ACCEPT_TOKEN(aux_sym_eof);
		END_STATE();
      }
```

So, instead of advancing to say state 4 and accepting `aux_sym_eof` there, just accept in place in state 0 instead

maxbrunsfeld commented on cli/npm/dsl.d.ts:
```diff
@@ -344,6 +345,20 @@ declare const token: {
+ * This is useful in languages where the top level grammar is a repeated
+ * sequence of statements followed by a terminator. One might use newlines
+ * for a terminator, but the end of the file also should match. Previously,
+ * '\0' was used to match eof, but this can match any NUL byte in the input,
```
I'd remove the `Previously, '\0' was used` sentence from the docs. I'm assuming you mean some specific grammar used `\0` to detect EOF? Either way, I don't think the docs need to talk about some thing that previously existed.

ItzCeeleeAdam commented:
i was completed

clason commented:
@amaanq is this still on the docket?

foxyseta commented:
> @amaanq is this still on the docket?

I was curious, too.

LostTime76 commented:
Bump. Update on this?

I have tokens that have potential to be terminated by a line terminator or the end of the file. Being able to use eof() within the grammar.js would be perfect. Otherwise, we have sort of a hack that every text needs to have a line terminator at the end. That especially gets really annoying for single line comments that should be allowed to terminate using a line terminator or EOF.

amaanq committed:
feat!: add an eof function

WillLillis committed:
fix(generate): model eof() as reduce-on-lookahead, not a shift target

This refines the eof() implementation so it doesn't explode the parse
state count.

Previously, `eof()` was lowered to `Symbol::end()` as a regular
production step. The parser would shift EOF onto the stack and then
reduce, which (a) required supporting plumbing for End in places that
previously never saw it, and (b) caused state-count regressions.

Instead, treat `requires_eof_lookahead` as a property of the
`Production`. When a flattened production ends in `eof()`, drop the
trailing End step and set the flag; emit the Reduce action only under
the end-of-input lookahead column.

WillLillis commented:
I rebased this and added a fix based on Max's feedback as a separate commit (to be squashed in if this is the right approach). 

- Productions ending in `eof()` are flattened with the end step stripped and a new `requires_eof_lookahead` flag set on the `Production`. The `Reduce` action is then emitted only under the EOF lookahead column instead of as a shift.

- `eof()` placed mid-sequence (i.e. inside an inlined `terminator = choice(..., eof())` used in a position where something must follow) is silently dropped at flatten time, since no input can follow EOF. If every alternative of a rule drops this way, generation fails with a clear `NoReachableProductions` error.
  - Two LR-table correctness fixes that turned out to be required:
    - The closure was propagating the parent's full FOLLOW set through requires_eof_lookahead productions, polluting inner reduce lookaheads with unrelated tokens. Restricting to {EOF} in the propagation eliminates a class of spurious conflicts.
    - ParseItem's Eq/Hash/Ord didn't include requires_eof_lookahead, so two productions differing only in that flag merged into one item and clobbered each other's production_id. Including it fixes correctness of reduce dispatch.


I tested with this change to tree-sitter-go:

```diff
diff --git a/grammar.js b/grammar.js
index 3648f02..b0315ea 100644
--- a/grammar.js
+++ b/grammar.js
@@ -26,7 +26,7 @@ const assignmentOperators = multiplicativeOperators.concat(additiveOperators).ma
 
 
 const newline = /\n/;
-const terminator = choice(newline, ';', '\0');
+const terminator = choice(newline, ';', eof());
 
 const hexDigit = /[0-9a-fA-F]/;
 const octalDigit = /[0-7]/;
@@ -115,15 +115,12 @@ module.exports = grammar({
   ],
 
   rules: {
-    source_file: $ => seq(
-      repeat(choice(
-        // Unlike a Go compiler, we accept statements at top-level to enable
-        // parsing of partial code snippets in documentation (see #63).
-        seq($._statement, terminator),
-        seq($._top_level_declaration, terminator),
-      )),
-      optional($._top_level_declaration),
-    ),
+    source_file: $ => repeat(seq(
+      // Unlike a Go compiler, we accept statements at top-level to enable
+      // parsing of partial code snippets in documentation (see #63).
+      choice($._statement, $._top_level_declaration),
+      terminator,
+    )),
 
     _top_level_declaration: $ => choice(
       $.package_clause,
```

All corpus tests still pass, and the state counts actually reduced a bit:
	- state count: 1442->1415
	- large state count: 29->28
	- symbol count: 214->213

WillLillis commented:
Slight wrinkle in the implementation, but shouldn't be blocking. I added an error for when a rule doesn't have any productions, as this can lead to silently broken parsers. The original intent behind this was for rules that dropped all of their eof productions and ended up empty. Unfortunately, tree-sitter-markdown actually uses an empty `choice` in its grammar today, which triggers the new error:

https://github.com/tree-sitter-grammars/tree-sitter-markdown/blob/c3570720f7f7bbad22fe96603f106276618e0cf5/tree-sitter-markdown/grammar.js#L27

Here's a minimal grammar demonstrating the problem:

```js
module.exports = grammar({
      name: 'empty_choice_demo',
      rules: {
          source_file: $ => seq('hello', $.dead, 'world'),
          dead: $ => choice(),
      }
  });
```

`tree-sitter generate` runs without any issues. However, if you try running `tree-sitter parse` on `hello world` you get a broken tree: `(ERROR [0, 0] - [0, 11])`.

I think this is worth fixing once I have a chance to understand tree-sitter-markdown's use case more, but shouldn't block this change. Instead (for now), we'll only emit an error if the rule _started_ with some choices and then flattening eliminated all of them.

WillLillis commented:
Rebased on top of latest master. Semantics should still be identical. Perf is neutral on walltime, at worst a 1% regression in peak rss for some grammars (e.g. razor).

clason commented:
This is one case where an error is actually appropriate ;) Emitting a broken parser would be the worse option.

Could you open an issue at tree-sitter-markdown about this for visibility? (Ideally, that kludge should be removed anyway.)

(Rebased the commit for good measure.)

WillLillis commented:
Amended the commit and noting here: The commit is marked as breaking because it changes some public types for `tree-sitter-generate`. The original implementation on this branch was also breaking for the C lib, but this is no longer the case.
